src/Common/src/CoreLib/System/IO/PathInternal.Windows.cs

```diff
@@ -188,64 +188,57 @@ internal static int GetRootLength(ReadOnlySpan<char> path)
         {
             int pathLength = path.Length;
             int i = 0;
-            int volumeSeparatorLength = 2;  // Length to the colon "C:"
-            int uncRootLength = 2;          // Length to the start of the server name "\\"
-            int devicePrefixLength = PathInternal.ExtendedPathPrefix.Length;
 
             bool deviceSyntax = IsDevice(path);
             bool deviceUnc = deviceSyntax && IsDeviceUNC(path);
-            if (deviceSyntax)
-            {
-                // Shift the position we look for the root from to account for the extended prefix
-                if (deviceUnc)
-                {
-                    // "\\" -> "\\?\UNC\"
-                    uncRootLength = UncExtendedPathPrefix.Length;
-                }
-                else if (devicePrefixLength + 1 < pathLength && path[devicePrefixLength + 1] == VolumeSeparatorChar && IsValidDriveChar(path[devicePrefixLength]))
-                {
-                    // "C:" -> "\\?\C:"
-                    volumeSeparatorLength += devicePrefixLength;
-                }
-            }
 
             if ((!deviceSyntax || deviceUnc) && pathLength > 0 && IsDirectorySeparator(path[0]))
             {
                 // UNC or simple rooted path (e.g. "\foo", NOT "\\?\C:\foo")
-
-                i = 1; //  Drive rooted (\foo) is one character
                 if (deviceUnc || (pathLength > 1 && IsDirectorySeparator(path[1])))
                 {
-                    // UNC (\\?\UNC\ or \\), scan past the next two directory separators at most
-                    // (e.g. to \\?\UNC\Server\Share or \\Server\Share\)
-                    i = uncRootLength;
-                    int n = 2; // Maximum separators to skip
+                    // UNC (\\?\UNC\ or \\), scan past server\share
+
+                    // Start past the prefix ("\\" or "\\?\UNC\")
+                    i = deviceUnc ? UncExtendedPrefixLength : UncPrefixLength;
+
+                    // Skip two separators at most
+                    int n = 2;
                     while (i < pathLength && (!IsDirectorySeparator(path[i]) || --n > 0))
```

Comment:
I find this hard to read as normally you would expect the termination check to be &&. But I see what you are doing is if the first clause is false, only then does it decrement n.

```c#
                     while (i < pathLength && (!IsDirectorySeparator(path[i]) || --n > 0))
                         i++;	                        
```
is this as good?
```c#
                     while (i < pathLength && n > 0))
                        if (IsDirectorySeparator(path[i]))
                                   n--;
                         i++;	                        
```

Reply:
Yeah, this one I've never been happy with. Some legacy logic I've brought along. I'll give it some thought.